syntax = "proto3";

import "messages.proto";
import "compatibility.proto";

service ServiceConnector {

    // This should always be run on app/client start.
    rpc CheckCompatibility(SemVer) returns (CompatibilityInfo) {}

    rpc Login(LoginRequest) returns (LoginResponse) {}

    rpc SendMessage(Message) returns (SendMessageResponse) {}
    rpc SubscribeToMessages(SubscribeToMessagesParams) returns (stream Message) {}
}

message SubscribeToMessagesParams {}

message LoginRequest {
    string username = 1;
    string password = 2;
    string two_factor_code = 3;
}

message LoginResponse {
    LoginStatus status = 1;
}

enum LoginStatus {
    LOGIN_OK = 0; // successfully logged in
    LOGIN_ERROR = 1; // unknown error

    LOGIN_WRONG_USERNAME = 2;
    LOGIN_WRONG_PASSWORD = 3;
    LOGIN_WRONG_2FA_CODE = 4;
    LOGIN_WRONG = 5;
    LOGIN_BANNED = 6;
    LOGIN_NEEDS_SECOND_FACTOR = 7;
}
